ddtrace/contrib/molten/patch.py

```diff
@@ -0,0 +1,180 @@
+import inspect
+import os
+
+import wrapt
+
+import molten
+from molten.http import Request
+
+from ... import Pin, config
+from ...ext import AppTypes, http
+from ...propagation.http import HTTPPropagator
+from ...utils.formats import asbool, get_env
+from ...utils.importlib import func_name
+from ...utils.wrappers import unwrap
+
+
+# Configure default configuration
+config._add('molten', dict(
+    service_name=get_env('molten', 'service_name', 'molten'),
+    app='molten',
+    app_type=AppTypes.web,
+    distributed_tracing_enabled=False,
+))
+
+def patch():
+    """Patch the instrumented methods
+    """
+    if getattr(molten, '_datadog_patch', False):
+        return
+    setattr(molten, '_datadog_patch', True)
+
+    pin = Pin(
+        service=config.molten['service_name'],
+        app=config.molten['app'],
+        app_type=config.molten['app_type'],
+    )
+
+    # add pin to module since many classes use __slots__
+    pin.onto(molten)
+
+    _w = wrapt.wrap_function_wrapper
+    _w('molten', 'App.__init__', patch_app_init)
+    _w('molten', 'App.__call__', patch_app_call)
+    _w('molten', 'Router.add_route', patch_add_route)
+
+def unpatch():
+    """Remove instrumentation
+    """
+    if getattr(molten, '_datadog_patch', False):
+        setattr(molten, '_datadog_patch', False)
+        unwrap(molten.App, '__init__')
+        unwrap(molten.App, '__call__')
+        unwrap(molten.Router, 'add_route')
+
+def trace_func(resource):
+    """Trace calls to function using provided resource name
+    """
+    @wrapt.function_wrapper
+    def _trace_func(wrapped, instance, args, kwargs):
+        pin = Pin.get_from(molten)
+
+        if not pin or not pin.enabled():
+            return wrapped(*args, **kwargs)
+
+        with pin.tracer.trace(func_name(wrapped), service=pin.service, resource=resource):
+            return wrapped(*args, **kwargs)
+
+    return _trace_func
+
+def trace_middleware(middleware):
+    """Trace calling of middleware function or object
+    """
+    @wrapt.function_wrapper
+    def _trace_middleware(wrapped, instance, args, kwargs):
+        pin = Pin.get_from(molten)
+        if not pin or not pin.enabled():
+            return wrapped(*args, **kwargs)
+
+        resource = None
+        if inspect.isfunction(wrapped):
```

Comment:
We will want to use `resource = func_name(wrapped)` here.

We can also just have `trace_func(resource=None)`, since by default if `tracer.trace(resource=None)` it'll inherit the resource name from the `span.name` property.